src/m3ninx/index/segment/builder/multi_segments_builder.go

```diff
@@ -147,6 +153,7 @@ func (b *builderFromSegments) AddSegments(segments []segment.Segment) error {
 			d := iter.Current()
 			negativeOffsets = append(negativeOffsets, currOffset)
 			if b.idSet.Contains(d.ID) {
+				b.reconcileDupeFn(d)
```

Comment:
I see yeah we kind of want to do patch up state if we filter out a duplicate.

I would probably update the DocumentsFilter interface to be three methods:
```go
type DocumentsFilter interface {
  ContainsDoc(d doc.Metadata) bool
  OnNotContainsDoc(d doc.Metadata)
  OnDuplicateDoc(d doc.Metadata)
}
```

That way we don't have to pass three different distinct things to the multi segments builder (the doc filter, the filter count and the reconcile duplicate function).

Then this becomes:
```go
		for iter.Next() {
			d := iter.Current()
			negativeOffsets = append(negativeOffsets, currOffset)
			if b.idSet.Contains(d.ID) {
				// Skip duplicates.
				negativeOffsets[len(negativeOffsets)-1] = -1
				currOffset++
				if b.filter != nil {
					b.filter.OnDuplicateDoc(d) // Callback here to play TryReconcileDuplicates
				}
				continue
			}
			if b.filter != nil && !b.filter.ContainsDoc(d) { // Existing call here to determine if should GC or not
				// Actively filtering and ID is not contained.
				negativeOffsets[len(negativeOffsets)-1] = -1
				currOffset++
				b.filter.OnNotContainsDoc(d) // Callback here to increment filter count
				continue
			}
			b.idSet.SetUnsafe(d.ID, struct{}{}, IDsMapSetUnsafeOptions{
				NoCopyKey:     true,
				NoFinalizeKey: true,
			})
			b.docs = append(b.docs, d)
			added++
		}
```

Reply:
What's the reason for passing in these functions to mutableSegments at all? IMO it might just be more clear for mutableSegments to call `d.Metadata.OnIndexSeries.Merge(...)` directly?

Reply:
Nice! I would be leaning towards Ryan's approach here; think the main upside of this `DocumentsFilter` approach would be getting metrics plumbed through, but we already have some through the `Entry` plumbing that's part of this PR; happy to refactor it to lean more heavily on the `DocumentsFilter` approach though if that can give us some benefit. Also needs a little annoying refactoring here: https://github.com/m3db/m3/blob/master/src/dbnode/storage/index/mutable_segments.go#L771 since we'd want to set a "partial" `DocumentsFilter` (not a huge change but still a bit annoying)